README.md
# 🗓 Calendar

Display Calendar with adding events for popular frameworks (Bootstrap, TailWinds...).

[![Latest Version on Packagist](https://img.shields.io/packagist/v/brilik/calendar.svg?style=flat-square)](https://packagist.org/packages/brilik/calendar)
[![Build Status](https://img.shields.io/travis/brilik/calendar/master.svg?style=flat-square)](https://travis-ci.org/brilik/calendar)
[![Quality Score](https://img.shields.io/scrutinizer/g/brilik/calendar.svg?style=flat-square)](https://scrutinizer-ci.com/g/brilik/calendar)
[![Total Downloads](https://img.shields.io/packagist/dt/brilik/calendar.svg?style=flat-square)](https://packagist.org/packages/brilik/calendar)

## Requirements

- PHP 7.3+

## Installation

You can install the package via composer:

```bash
composer require brilik/calendar
```

## How it works

```php
use VitoBryliano\Calendar\TailWindCalendar;
use VitoBryliano\Calendar\BootstrapCalendar;

// Get date in ISO format
$date = Carbon::now()->tz('Europe/Kiev')->format('Y-m-d');
// You can create calendar in the Bootstrap styles
$calendar = new BootstrapCalendar($date);
// Or you can create calendar in the TailWind styles
$calendar = new TailWindCalendar($date);
// You can add navigate for choose month
$calendar->navigation = true;
// You can add custom name months or their translations
$calendar->setWeekDaysName([
    __('Mon'),
    __('Tue'),
    __('Wed'),
    __('Thu'),
    __('Fri'),
    __('Sut'),
    __('Sun')
]);
```

### Changelog

Please see [CHANGELOG](CHANGELOG.md) for more information what has changed recently.

## License

The MIT License (MIT). Please see [License File](LICENSE) for more information.


CHANGELOG.md
# Changelog

All notable changes to `calendar` will be documented in this file

## 1.0.0 - 2023-08-06

- Added Bootstrap and TailWind styles to the Calendar


src/TailWindCalendar.php
<?php

namespace VitoBryliano\Calendar;

use VitoBryliano\Calendar\components\modal\ButtonOpenModalComponent;
use VitoBryliano\Calendar\components\NavigationComponent;
use VitoBryliano\Calendar\interface\ICalendar;
use VitoBryliano\Calendar\traits\ActionClickPopover;
use VitoBryliano\Calendar\traits\WeekDaysNamePanelTrait;
use Breadthe\SimpleCalendar\Calendar;
use Carbon\Carbon;

/**
 * todo Next version:
 *  1. Need add translate for month and for week days name (Carbon::setLocale('uk_UA');).
 */
final class TailWindCalendar extends Calendar implements ICalendar
{
    use ActionClickPopover, WeekDaysNamePanelTrait;

    public bool $navigation = false;
    private string $table;
    private readonly Carbon $currentDate;

    /**
     * @param string $today ISO date
     */
    public function __construct(string $today)
    {
        parent::__construct($today);

        $this->currentDate = Carbon::parse($today);
    }

    /** Return HTML of the Calendar.
     * @return string
     */
    public function render(): string
    {
        $navigation = $this->navigation ? $this->getNavigationHtml() : '';
        $popover = $this->popoverByClickDay ? $this->getModalHtml('My custom click') : '';

        return $navigation . $this->getTableHtml() . $popover;
    }

    /** Return HTML of table.
     * @return string
     */
    private function getTableHtml(): string
    {
        $table = '
        <table class="w-full text-sm text-left text-gray-500 dark:text-gray-400">
            <thead class="text-xs text-gray-700 uppercase bg-gray-50 dark:bg-gray-700 dark:text-gray-400">
            <tr>';

        foreach ($this->weekDaysName as $day) {
            $table .= '<th scope="col" class="px-6 py-3">' . $day . '</th>';
        }

        $table .= '</tr></thead><tbody>';
        $table .= '<tr class="bg-white border-b dark:bg-gray-800 dark:border-gray-700">';
        $prevDate = null;

        foreach ($this->grid() as $date):
            // init of variables for current month
            $btn = new ButtonOpenModalComponent();
            $btn->setClass('text-gray-900 hover:text-white border border-gray-800 hover:bg-gray-900 focus:ring-4 focus:outline-none focus:ring-gray-300 font-medium rounded-lg text-sm px-5 py-2.5 text-center mr-2 mb-2 dark:border-gray-600 dark:text-gray-400 dark:hover:text-white dark:hover:bg-gray-600 dark:focus:ring-gray-800')
                ->setBtnContent($date->day);
            // if item date equals today
            if ($date->eq($this->currentDate)) {
                $btn->setClass('text-white bg-gray-800 hover:bg-gray-900 focus:outline-none focus:ring-4 focus:ring-gray-300 font-medium rounded-lg text-sm px-5 py-2.5 mr-2 mb-2 dark:bg-gray-800 dark:hover:bg-gray-700 dark:focus:ring-gray-700 dark:border-gray-700');
            }
            // if isn't current month
            if ($date->month < $this->currentDate->month || $date->month > $this->currentDate->month) {
                $btn->setClass('p-3 w-full text-gray-300')
                    ->disabled('cursor-not-allowed');
            } else {
                // if is end week
                if ($date->weekNumberInMonth > $prevDate->weekNumberInMonth) {
                    $table .= '<tr class="bg-white border-b dark:bg-gray-800 dark:border-gray-700"></tr>';
                }
            }

            $prevDate = $date;
            $table .= '<td class="p-4 text-center">' . $this->getButtonOpenModal($btn) . '</td>';
        endforeach;

        $table .= '</tr>';
        $table .= '</tbody></table>';

        return $table;
    }

    /** Return HTML of navigation.
     * @return string
     */
    private function getNavigationHtml(): string
    {
        return (new NavigationComponent())
            ->setNavClass('pt-4')
            ->setNavAriaLabel('Table navigation')
            ->setListClass('flex items-center justify-between py-5')
            ->setBtnLeftHref($this->startOfPrevMonth->format('Y-m-d'))
            ->setBtnLeftClass('font-medium text-gray-600 dark:text-gray-500 hover:underline')
            ->setBtnLeftContent($this->startOfPrevMonth->format('F'))
            ->setBtnRightHref($this->startOfNextMonth->format('Y-m-d'))
            ->setBtnRightClass('font-medium text-gray-600 dark:text-gray-500 hover:underline')
            ->setBtnRightContent($this->startOfNextMonth->format('F'))
            ->setTitleClass('text-3xl text-gray-900 dark:text-white')
            ->setTitleContent($this->currentDate->format('F Y'))
            ->toString();
    }
}


src/traits/ActionClickPopover.php
<?php
/*
 * Copyright (c) 2023. Lorem ipsum dolor sit amet, consectetur adipiscing elit.
 * Morbi non lorem porttitor neque feugiat blandit. Ut vitae ipsum eget quam lacinia accumsan.
 * Etiam sed turpis ac ipsum condimentum fringilla. Maecenas magna.
 * Proin dapibus sapien vel ante. Aliquam erat volutpat. Pellentesque sagittis ligula eget metus.
 * Vestibulum commodo. Ut rhoncus gravida arcu.
 */

namespace VitoBryliano\Calendar\traits;

use VitoBryliano\Calendar\components\modal\ButtonOpenModalComponent;

/**
 * todo Next version:
 *  1. Modal content is duplicate, so u need add content using date in setModalContent(date, html).
 *  2. Use DTO for set of attributes button and modal.
 */
trait ActionClickPopover
{
    private string $modalContent = '<p>And here\'s some amazing content. It\'s very engaging. Right?</p>';
    private ButtonOpenModalComponent $button;
    /** todo Next version:
     *   1. Need to be able to add events to the calendar.
     * @var bool
     */
    public bool $popoverByClickDay = false;

    /** Set HTML content in the modal.
     *
     * @param  string  $html
     *
     * @return void
     */
    public function setModalContent(string $html = ''): void
    {
        $this->modalContent = $html;
    }

    /** Get HTML button which open modal.
     *
     * @param  ButtonOpenModalComponent  $button
     *
     * @return string
     */
    private function getButtonOpenModal(ButtonOpenModalComponent $button): string
    {
        return $button->toString();
    }

    /** Return HTML of modal. Requires bootstrap.bundle.min.js for it.
     *  todo Next version:
     *   1. Need use the components/modal/ModalContentComponent for set attributes in HTML.
     * @param  string  $title
     *
     * @return string
     * @requires bootstrap.bundle.min.js
     */
    private function getModalHtml(string $title = 'Popover click'): string
    {
        return '<div data-popover id="popover-click" role="tooltip" class="absolute z-10 invisible inline-block w-64 text-sm text-gray-500 transition-opacity duration-300 bg-white border border-gray-200 rounded-lg shadow-sm opacity-0 dark:text-gray-400 dark:border-gray-600 dark:bg-gray-800">
                    <div class="px-3 py-2 bg-gray-100 border-b border-gray-200 rounded-t-lg dark:border-gray-600 dark:bg-gray-700">
                        <h3 class="font-semibold text-gray-900 dark:text-white">' . $title . '</h3>
                    </div>
                    <div class="px-3 py-2">' . $this->modalContent . '</div>
                    <div data-popper-arrow></div>
                </div>';
    }
}


src/components/NavigationComponent.php
<?php

namespace VitoBryliano\Calendar\components;

class NavigationComponent
{
    private string $class = '';
    private string $ariaLabel = '';
    private string $listClass = '';
    private string $btnLeftHref = '';
    private string $btnLeftClass = '';
    private string $btnLeftContent = '';
    private string $btnRightHref = '';
    private string $btnRightClass = '';
    private string $btnRightContent = '';
    private string $titleClass = '';
    private string $titleContent = '';

    /**
     * @param string $string
     * @return $this
     */
    public function setNavClass(string $string): self
    {
        $this->class = $string;

        return $this;
    }

    /**
     * @param string $ariaLabel
     * @return NavigationComponent
     */
    public function setNavAriaLabel(string $ariaLabel): self
    {
        $this->ariaLabel = $ariaLabel;

        return $this;
    }

    /**
     * @param string $listClass
     * @return NavigationComponent
     */
    public function setListClass(string $listClass): self
    {
        $this->listClass = $listClass;

        return $this;
    }

    /**
     * @param string $btnLeftHref
     * @return NavigationComponent
     */
    public function setBtnLeftHref(string $btnLeftHref): self
    {
        $this->btnLeftHref = $btnLeftHref;

        return $this;
    }

    /**
     * @param string $btnLeftClass
     * @return NavigationComponent
     */
    public function setBtnLeftClass(string $btnLeftClass): self
    {
        $this->btnLeftClass = $btnLeftClass;

        return $this;
    }

    /**
     * @param string $btnLeftContent
     * @return NavigationComponent
     */
    public function setBtnLeftContent(string $btnLeftContent): self
    {
        $this->btnLeftContent = $btnLeftContent;

        return $this;
    }

    /**
     * @param string $btnRightHref
     * @return NavigationComponent
     */
    public function setBtnRightHref(string $btnRightHref): self
    {
        $this->btnRightHref = $btnRightHref;

        return $this;
    }

    /**
     * @param string $btnRightClass
     * @return NavigationComponent
     */
    public function setBtnRightClass(string $btnRightClass): self
    {
        $this->btnRightClass = $btnRightClass;

        return $this;
    }

    /**
     * @param string $btnRightContent
     * @return NavigationComponent
     */
    public function setBtnRightContent(string $btnRightContent): self
    {
        $this->btnRightContent = $btnRightContent;

        return $this;
    }

    /**
     * @param string $string
     * @return $this
     */
    public function setTitleClass(string $string): self
    {
        $this->titleClass = $string;

        return $this;
    }

    /**
     * @param string $titleContent
     * @return NavigationComponent
     */
    public function setTitleContent(string $titleContent): self
    {
        $this->titleContent = $titleContent;

        return $this;
    }

    public function toString(): string
    {
        return '<nav class="' . $this->class . '" aria-label="' . $this->ariaLabel . '">
            <ul class="' . $this->listClass . '">
                <li>
                    <a href="?date=' . $this->btnLeftHref . '" class="' . $this->btnLeftClass . '" role="button">
                     <span aria-hidden="true">&laquo;</span>&nbsp;' . $this->btnLeftContent . '</a>
                </li>
                <li>
                    <span class="' . $this->titleClass . '">' . $this->titleContent . '</span>
                </li>
                <li>
                    <a href="?date=' . $this->btnRightHref . '" class="' . $this->btnRightClass . '" role="button">
                    ' . $this->btnRightContent . '&nbsp;<span aria-hidden="true" >&raquo;</span></a>
                </li>
            </ul>
        </nav>';
    }
}


src/components/modal/ButtonOpenModalComponent.php
<?php

namespace VitoBryliano\Calendar\components\modal;

class ButtonOpenModalComponent
{
    private string $class;
    private string $btnContent;
    private string $disabled = '';

    /**
     * @param string $class
     * @return ButtonOpenModalComponent
     */
    public function setClass(string $class): self
    {
        $this->class = $class;

        return $this;
    }

    /**
     * @param string $btnContent
     * @return ButtonOpenModalComponent
     */
    public function setBtnContent(string $btnContent): self
    {
        $this->btnContent = $btnContent;

        return $this;
    }

    /**
     * @param string $class
     * @return ButtonOpenModalComponent
     */
    public function disabled(string $class = 'disabled'): self
    {
        $this->disabled = ' ' . $class;

        return $this;
    }

    public function toString(): string
    {
        $disabled = $this->disabled ? 'disabled' : '';

        return '<button type="button"
                        data-popover-target="popover-click"
                        data-popover-trigger="click"
                        ' . $disabled . '
                        class="' . $this->class . $this->disabled . '">' . $this->btnContent . '</button>';
    }
}


src/BootstrapCalendar.php
<?php

namespace VitoBryliano\Calendar;

use VitoBryliano\Calendar\components\modal\ButtonOpenModalComponent;
use VitoBryliano\Calendar\components\NavigationComponent;
use VitoBryliano\Calendar\interface\ICalendar;
use VitoBryliano\Calendar\traits\ActionClickPopover;
use VitoBryliano\Calendar\traits\WeekDaysNamePanelTrait;
use Breadthe\SimpleCalendar\Calendar;
use Carbon\Carbon;

final class BootstrapCalendar extends Calendar implements ICalendar
{
    use ActionClickPopover, WeekDaysNamePanelTrait;

    public bool $navigation = false;
    private string $table;
    private readonly Carbon $currentDate;

    /**
     * @param string $today ISO date
     */
    public function __construct(string $today)
    {
        parent::__construct($today);

        $this->currentDate = Carbon::parse($today);
    }

    /** Return HTML of table.
     * @return string
     */
    public function render(): string
    {
        $navigation = $this->navigation ? $this->getNavigationHtml() : '';
        $popover = $this->popoverByClickDay ? $this->getModalHtml('My custom click') : '';
        $script = $this->popoverByClickDay ? '<script>
        const needleScript = document.querySelector("script[src*=\'bootstrap.bundle.js\']")
        // if (needleScript) {
            // const script = document.createElement("script")
            // script.innerHTML =
            setTimeout(() => {
                console.log(\'asd\')
                const popover = new bootstrap.Popover(document.querySelector("#popover-click"), {
                    trigger: "focus"
                })
            }, 700)
            // document.head.appendChild(script)
        // }
        </script>' : '';

        return $navigation . $this->getTableHtml() . $popover . $script;
    }

    /** Return HTML of table.
     * @return string
     */
    private function getTableHtml(): string
    {
        $table = '
        <table class="table" style="max-width: 650px">
            <thead>
            <tr>';

        foreach ($this->weekDaysName as $day) {
            $table .= '<th scope="col" class="text-center">' . $day . '</th>';
        }

        $table .= '</tr></thead><tbody>';
        $table .= '<tr>';
        $prevDate = null;

        foreach ($this->grid() as $date):
            // init of variables for current month
            $btn = new ButtonOpenModalComponent();
            $btn->setClass('btn btn-outline-secondary p-3 w-100')
                ->setBtnContent($date->day);
            // if item date equals today
            if ($date->eq($this->currentDate)) {
                $btn->setClass('btn btn-primary p-3 w-100');
            }
            // if isn't current month
            if ($date->month < $this->currentDate->month || $date->month > $this->currentDate->month) {
                $btn->setClass('btn btn-outline-secondary p-3 w-100')
                    ->disabled();
            } else {
                // if is end week
                if ($date->weekNumberInMonth > $prevDate->weekNumberInMonth) {
                    $table .= '<tr></tr>';
                }
            }

            $prevDate = $date;
            $table .= '<td class="text-center">' . $this->getButtonOpenModal($btn) . '</td>';
        endforeach;

        $table .= '</tr>';
        $table .= '</tbody></table>';

        return $table;
    }

    /** Return HTML of navigation.
     * @return string
     */
    private function getNavigationHtml(): string
    {
        return (new NavigationComponent())
            ->setNavClass('w-100')
            ->setNavAriaLabel('Table navigation')
            ->setListClass('pagination align-items-center justify-content-evenly')
            ->setBtnLeftHref($this->startOfPrevMonth->format('Y-m-d'))
            ->setBtnLeftClass('btn btn-link')
            ->setBtnLeftContent($this->startOfPrevMonth->format('F'))
            ->setBtnRightHref($this->startOfNextMonth->format('Y-m-d'))
            ->setBtnRightClass('btn btn-link')
            ->setBtnRightContent($this->startOfNextMonth->format('F'))
            ->setTitleClass('text-3xl text-gray-900 dark:text-white')
            ->setTitleContent($this->currentDate->format('F Y'))
            ->toString();
    }
}
